File: src/at/fhooe/mc/emg/util/Configuration.java
package at.fhooe.mc.emg.util;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class Configuration {

	private static Configuration instance;

	private static final String KEY_WRITE_LOG = "write_log";
	private static final String KEY_COPY_TO_SIM = "copy_to_simulation";
	private static final String KEY_SIM_ENDLESS_LOOP = "simulation_endless_loop";
	private static final String KEY_RAVG_WS = "ravg_window_size";
	private static final String KEY_SG_WIDTH = "savitzky_golay_width";

	private static final String CONFIG_PATH = System.getProperty("user.dir") + "/data/config.json";

	private boolean isWriteToLogEnabled;
	private boolean isCopyToSimulationEnabled;
	private boolean isSimEndlessLoopEnabled;
	
	private int runningAverageWindowSize;
	private int savitzkyGolayFilterWidth;

	public static Configuration getInstance() {

		if (instance != null) {
			return instance;
		}
		instance = new Configuration();
		return instance;
	}

	public Configuration() {
		load();
	}

	private void load() {

		File file = new File(CONFIG_PATH);
		if (!file.exists()) {
			try {
				file.createNewFile();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

		try (BufferedReader reader = new BufferedReader(new FileReader(file))) {

			String line;
			StringBuilder buffer = new StringBuilder();
			while ((line = reader.readLine()) != null) {
				buffer.append(line);
			}

			JsonObject object = new JsonParser().parse(buffer.toString()).getAsJsonObject();
			isWriteToLogEnabled = (object.has(KEY_WRITE_LOG)) && object.get(KEY_WRITE_LOG).getAsBoolean();
			isCopyToSimulationEnabled = (object.has(KEY_COPY_TO_SIM)) && object.get(KEY_COPY_TO_SIM).getAsBoolean();
			isSimEndlessLoopEnabled = (object.has(KEY_SIM_ENDLESS_LOOP)) && object.get(KEY_SIM_ENDLESS_LOOP).getAsBoolean();
			runningAverageWindowSize = (object.has(KEY_RAVG_WS))
					? object.get(KEY_RAVG_WS).getAsInt() : 30;
			savitzkyGolayFilterWidth = (object.has(KEY_SG_WIDTH))
					? object.get(KEY_SG_WIDTH).getAsInt() : 10;

		} catch (Exception e) {
			e.printStackTrace();
			isWriteToLogEnabled = false;
			isCopyToSimulationEnabled = false;
			isSimEndlessLoopEnabled = false;
			runningAverageWindowSize = 30;
			savitzkyGolayFilterWidth = 10;
		}

	}

	public void save() {

		JsonObject object = new JsonObject();
		object.addProperty(KEY_WRITE_LOG, isWriteToLogEnabled);
		object.addProperty(KEY_COPY_TO_SIM, isCopyToSimulationEnabled);
		object.addProperty(KEY_SIM_ENDLESS_LOOP, isSimEndlessLoopEnabled);
		object.addProperty(KEY_RAVG_WS, runningAverageWindowSize);
		object.addProperty(KEY_SG_WIDTH, savitzkyGolayFilterWidth);

		try {
			AppUtils.writeFile(new File(CONFIG_PATH), object.toString());
		} catch (IOException e) {
			e.printStackTrace();
		}

	}

	public boolean isWriteToLogEnabled() {
		return isWriteToLogEnabled;
	}

	public void setWriteToLogEnabled(boolean isWriteToLogEnabled) {
		this.isWriteToLogEnabled = isWriteToLogEnabled;
	}

	public boolean isCopyToSimulationEnabled() {
		return isCopyToSimulationEnabled;
	}

	public void setCopyToSimulationEnabled(boolean isCopyToSimulationEnabled) {
		this.isCopyToSimulationEnabled = isCopyToSimulationEnabled;
	}
	
	public boolean isSimulationEndlessLoopEnabled() {
		return isSimEndlessLoopEnabled;
	}
	
	public void setSimulationEndlessLoopEnabled(boolean isSimEndlessLoopEnabled) {
		this.isSimEndlessLoopEnabled = isSimEndlessLoopEnabled;
	}

	public int getRunningAverageWindowSize() {
		return runningAverageWindowSize;
	}

	public void setRunningAverageWindowSize(int runningAverageWindowSize) {
		this.runningAverageWindowSize = runningAverageWindowSize;
	}

	public int getSavitzkyGolayFilterWidth() {
		return savitzkyGolayFilterWidth;
	}

	public void setSavitzkyGolayFilterWidth(int savitzkyGolayFilterWidth) {
		this.savitzkyGolayFilterWidth = savitzkyGolayFilterWidth;
	}

}

File: core/src/at/fhooe/mc/emg/core/util/Configuration.kt
package at.fhooe.mc.emg.core.util

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.stream.Collectors

object Configuration {

    private val KEY_WRITE_LOG = "write_log"
    private val KEY_COPY_TO_SIM = "copy_to_simulation"
    private val KEY_SIM_ENDLESS_LOOP = "simulation_endless_loop"
    private val KEY_RAVG_WS = "ravg_window_size"
    private val KEY_SG_WIDTH = "savitzky_golay_width"

    private val CONFIG_PATH = System.getProperty("user.dir") + "/data/config.json"

    var isWriteToLogEnabled: Boolean = false
    var isCopyToSimulationEnabled: Boolean = false
    var isSimulationEndlessLoopEnabled: Boolean = false

    var runningAverageWindowSize: Int = 0
    var savitzkyGolayFilterWidth: Int = 0

    init {
        load()
    }

    private fun load() {

        try {

            val file = File(CONFIG_PATH)
            if (!file.exists()) {
                file.createNewFile()
            }

            val strJson = Files.readAllLines(Paths.get(file.absolutePath))
                    .stream().collect(Collectors.joining("\n"))

            val element = JsonParser().parse(strJson).asJsonObject
            isWriteToLogEnabled = element.has(KEY_WRITE_LOG) && element.get(KEY_WRITE_LOG).asBoolean
            isCopyToSimulationEnabled = element.has(KEY_COPY_TO_SIM) && element.get(KEY_COPY_TO_SIM).asBoolean
            isSimulationEndlessLoopEnabled = element.has(KEY_SIM_ENDLESS_LOOP) && element.get(KEY_SIM_ENDLESS_LOOP).asBoolean
            runningAverageWindowSize = if (element.has(KEY_RAVG_WS)) element.get(KEY_RAVG_WS).asInt else 30
            savitzkyGolayFilterWidth = if (element.has(KEY_SG_WIDTH)) element.get(KEY_SG_WIDTH).asInt else 10

        } catch (e: Exception) {
            e.printStackTrace()
            isWriteToLogEnabled = false
            isCopyToSimulationEnabled = false
            isSimulationEndlessLoopEnabled = false
            runningAverageWindowSize = 30
            savitzkyGolayFilterWidth = 10
        }
    }

    fun save() {

        val element = JsonObject()
        element.addProperty(KEY_WRITE_LOG, isWriteToLogEnabled)
        element.addProperty(KEY_COPY_TO_SIM, isCopyToSimulationEnabled)
        element.addProperty(KEY_SIM_ENDLESS_LOOP, isSimulationEndlessLoopEnabled)
        element.addProperty(KEY_RAVG_WS, runningAverageWindowSize)
        element.addProperty(KEY_SG_WIDTH, savitzkyGolayFilterWidth)

        try {
            AppUtils.writeFile(File(CONFIG_PATH), element.toString())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

}
